File: src/brayns/core/engine/Texture.cpp
/* Copyright (c) 2015-2024 EPFL/Blue Brain Project
 * All rights reserved. Do not distribute without permission.
 *
 * Responsible Author: adrien.fleury@epfl.ch
 *
 * This file is part of Brayns <https://github.com/BlueBrain/Brayns>
 *
 * This library is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License version 3.0 as published
 * by the Free Software Foundation.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this library; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

#include "Texture.h"

namespace
{
using namespace brayns;

void setTexture2DParams(OSPTexture handle, const Texture2DSettings &settings)
{
    setObjectParam(handle, "filter", static_cast<OSPTextureFilter>(settings.filter));
    setObjectParam(handle, "wrapMode", static_cast<OSPTextureWrapMode>(settings.wrap));
}

void setVolumeTextureParams(OSPTexture handle, const VolumeTextureSettings &settings)
{
    setObjectParam(handle, "volume", settings.volume);
    setObjectParam(handle, "transferFunction", settings.transferFunction);
}
}

namespace brayns
{
DataType getDataType(TextureFormat format)
{
    switch (format)
    {
    case TextureFormat::Rgba8:
        return OSP_VEC4UC;
    case TextureFormat::Srgba8:
        return OSP_VEC4UC;
    case TextureFormat::Rgba32F:
        return OSP_VEC4F;
    case TextureFormat::Rgb8:
        return OSP_VEC3UC;
    case TextureFormat::Srgb8:
        return OSP_VEC3UC;
    case TextureFormat::Rgb32F:
        return OSP_VEC3F;
    case TextureFormat::R8:
        return OSP_UCHAR;
    case TextureFormat::Ra8:
        return OSP_VEC2UC;
    case TextureFormat::L8:
        return OSP_UCHAR;
    case TextureFormat::La8:
        return OSP_VEC2UC;
    case TextureFormat::R32F:
        return OSP_FLOAT;
    case TextureFormat::Rgba16:
        return OSP_VEC4US;
    case TextureFormat::Rgb16:
        return OSP_VEC3US;
    case TextureFormat::Ra16:
        return OSP_VEC2US;
    case TextureFormat::R16:
        return OSP_USHORT;
    default:
        throw std::invalid_argument("Invalid texture format");
    }
}

std::size_t getSize(TextureFormat format)
{
    switch (format)
    {
    case TextureFormat::Rgba8:
        return 4;
    case TextureFormat::Srgba8:
        return 4;
    case TextureFormat::Rgba32F:
        return 16;
    case TextureFormat::Rgb8:
        return 3;
    case TextureFormat::Srgb8:
        return 3;
    case TextureFormat::Rgb32F:
        return 12;
    case TextureFormat::R8:
        return 1;
    case TextureFormat::Ra8:
        return 2;
    case TextureFormat::L8:
        return 1;
    case TextureFormat::La8:
        return 2;
    case TextureFormat::R32F:
        return 4;
    case TextureFormat::Rgba16:
        return 8;
    case TextureFormat::Rgb16:
        return 6;
    case TextureFormat::Ra16:
        return 4;
    case TextureFormat::R16:
        return 2;
    default:
        throw std::invalid_argument("Invalid texture format");
    }
}

void Texture2D::update(const Texture2DSettings &settings)
{
    auto handle = getHandle();
    setTexture2DParams(handle, settings);
    commitObject(handle);
}

Texture2D createTexture2D(Device &device, const TextureData2D &data, const Texture2DSettings &settings)
{
    auto handle = ospNewTexture("texture2d");
    auto texture = wrapObjectHandleAs<Texture2D>(device, handle);

    auto type = getDataType(data.format);
    auto wrapper = createData(data.value.data(), type, Size3(data.size, 1));

    setObjectParam(handle, "data", wrapper);
    setObjectParam(handle, "format", static_cast<OSPTextureFormat>(data.format));

    setTexture2DParams(handle, settings);

    commitObject(device, handle);

    return texture;
}

void VolumeTexture::update(const VolumeTextureSettings &settings)
{
    auto handle = getHandle();
    setVolumeTextureParams(handle, settings);
    commitObject(handle);
}

VolumeTexture createVolumeTexture(Device &device, const VolumeTextureSettings &settings)
{
    auto handle = ospNewTexture("volume");
    auto texture = wrapObjectHandleAs<VolumeTexture>(device, handle);

    setVolumeTextureParams(handle, settings);

    commitObject(device, handle);

    return texture;
}
}

File: python/brayns/api/core/texture.py
# Copyright (c) 2015-2024 EPFL/Blue Brain Project
# All rights reserved. Do not distribute without permission.
#
# Responsible Author: adrien.fleury@epfl.ch
#
# This file is part of Brayns <https://github.com/BlueBrain/Brayns>
#
# This library is free software; you can redistribute it and/or modify it under
# the terms of the GNU Lesser General Public License version 3.0 as published
# by the Free Software Foundation.
#
# This library is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more
# details.
#
# You should have received a copy of the GNU Lesser General Public License
# along with this library; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.

from dataclasses import dataclass
from enum import Enum
from typing import TypedDict, Unpack

from brayns.network.connection import Connection
from brayns.utils.composing import serialize
from brayns.utils.parsing import deserialize

from .objects import CreateObjectParams, Object, create_specific_object, get_specific_object, update_specific_object
from .transfer_function import TransferFunction
from .volume import Volume


class Texture(Object): ...


class Texture2D(Texture): ...


class TextureFormat(Enum):
    RGBA8 = "Rgba8"
    SRGBA8 = "Srgba8"
    RGBA32F = "Rgba32F"
    RGB8 = "Rgb8"
    SRGB8 = "Srgb8"
    RGB32F = "Rgb32F"
    R8 = "R8"
    RA8 = "Ra8"
    L8 = "L8"
    LA8 = "La8"
    R32F = "R32F"
    RGBA16 = "Rgba16"
    RGB16 = "Rgb16"
    RA16 = "Ra16"
    R16 = "R16"


class TextureFilter(Enum):
    LINEAR = "Linear"
    NEAREST = "Nearest"


class TextureWrap(Enum):
    REPEAT = "Repeat"
    MORRORED_REPEAT = "MirroredRepeat"
    CLAMP_TO_EDGE = "ClampToEdge"


@dataclass
class GetTexture2DResult:
    format: TextureFormat
    size: tuple[int, int]
    filter: TextureFilter
    wrap: TextureWrap


class UpdateTexture2DParams(TypedDict, total=False):
    filter: TextureFilter
    wrap: TextureWrap


class CreateTexture2DParams(CreateObjectParams, UpdateTexture2DParams):
    format: TextureFormat
    size: tuple[int, int]


async def create_texture2d(
    connection: Connection, data: bytes, **settings: Unpack[CreateTexture2DParams]
) -> Texture2D:
    object = await create_specific_object(connection, "Texture2D", serialize(settings), data)
    return Texture2D(object.id)


async def get_texture2d(connection: Connection, texture: Texture2D) -> GetTexture2DResult:
    result = await get_specific_object(connection, "Texture2D", texture)
    return deserialize(result, GetTexture2DResult)


async def update_texture2d(
    connection: Connection, texture: Texture2D, **settings: Unpack[UpdateTexture2DParams]
) -> None:
    await update_specific_object(connection, "Texture2D", texture, serialize(settings))


class VolumeTexture(Texture): ...


@dataclass
class GetVolumeTextureResult:
    volume: Volume
    transfer_function: TransferFunction


class UpdateVolumeTextureParams(TypedDict):
    volume: Volume
    transfer_function: TransferFunction


class CreateVolumeTextureParams(CreateObjectParams, UpdateVolumeTextureParams): ...


async def create_volume_texture(
    connection: Connection, **settings: Unpack[CreateVolumeTextureParams]
) -> VolumeTexture:
    object = await create_specific_object(connection, "VolumeTexture", serialize(settings))
    return VolumeTexture(object.id)


async def get_volume_texture(connection: Connection, texture: VolumeTexture) -> GetVolumeTextureResult:
    result = await get_specific_object(connection, "VolumeTexture", texture)
    return deserialize(result, GetVolumeTextureResult)


async def update_volume_texture(
    connection: Connection, texture: VolumeTexture, **settings: Unpack[UpdateVolumeTextureParams]
) -> None:
    await update_specific_object(connection, "VolumeTexture", texture, serialize(settings))
